[Documentation] a11y Markup Checklist (#558)
massgov/mayflower: patternlab/docs/a11y-checklist/headings.md

[⬅️ a11y Checklist](a11y-checklist.md)

# Headings `<h#>`

## Heading level

> A heading needs its **proper level number** to represent the hierarchy of the content in the relationship of the parent and child components.

Reviewer: ```suggestion
A heading's level number must accurately represent the hierarchy of the content in the relationship of the parent and child components.
```

> HTML5 allows you to start with `<h1>` in every `<section>`. That is valid as HTML5, but it doesn't help assistive technology users to grasp the content hierarchy in the page.

Reviewer: If this doesn't help assistive technology users, what is your recommendation? Do you want to refer to some other section here (e.g. to the part explaining that only the page title should use `<h1>`?)

Author: We reserves `<h1>` for a page title, which is only used once in a page. The rest of headings in the page should be defined in the relation of the page title `<h1>` because they are supposed to fall under the page thus they are `<h2>`, `<h3>` and so on although some utility type components might not make sense to this rule.  There might be multiple components in the same level in a page, so the page can have multiple `<h2>`s, `<h3>`s and so forth.

The headings to present the content structure of the  page should be defined using the only `<h1>` as their structural starting point.

### ➤➤ Is the heading level number flexible?

☑️ It can receive and consume the level number value from the parent component.

> ☑️ It can have proper level number depending on where the component is placed in a page in the sequential manner from small to large. (If its parent component has *h3*, this component starts with *h4*.)

Reviewer: ```suggestion
☑️ The component's heading level number should adjust according to the component's placement on a page. For example, if its parent component starts with an *h3*, this component starts with *h4*.
```

> The *sequential manner* means:

Reviewer: ```suggestion
Proper heading level number means:
```

> ☑️ No skipping - Some in-between number(s) is(are) NOT missing.

Reviewer: ```suggestion
☑️ No skipping - no in-between number(s) is(are) missing.
```

> ☑️ No jumping - NOT putting heading level numbers randomly.

Reviewer: ```suggestion
☑️ No jumping - heading level numbers always go from small to large (never large to small).
```

<!--
- [ ] No backward - Numbers does NOT go down from large to small.
-->

**➣ Yes:** The heading level number is set dynamically, and meets these conditions.

> **🛑 Watch out when your template contains multiple headings.** The heading levels could change in the template. You might need to adjust the level in the template.

Reviewer: I'm not sure what this means - can you explain it more?

Author: Sure.  This sometimes happens in organism level templates.  The template has a heading for its own.  It's not necessarily hardcoded, it might calls a heading template in it.  After that the template calls some child component templates, which contain their headings. Thus, one template contains multiple headings.  

The child components' heading levels are different from one for the heading for the initial template. So, the heading level value adjustment needs to happen between the heading for the initial template and the child components' in the initial template, not in its child component templates.

**Example:  03-organisms/by-author/about-section.twig** (note: the current heading levels are not correct in Mayflower.)

At L. 4, it calls **@atoms/04-headings/comp-heading.twig** for a heading for the about section as **\<h2\>** for example.

At L. 22, it calls **@organisms/by-author/person-card.twig** which contains a heading for person card. Since this is a child component of the about section, its heading should be **\<h3\>** by increasing its level by 1.  
This adjustment of the _level value_ should happen in **about-section.twig** (= parent component), not in its child component **@organisms/by-author/person-card.twig**, then the value is passed to the child component because the adjustment is only necessary in this context of the use of the 2 templates as the parent-child relationship. 

If this child component is used independently somewhere else and it increases its heading level value by 1 in the template, then its heading level could skip when another parent template with a different context from above passes its heading level value.

**Other samples:**

- 03-organisms/by-author/collapsible-content.twig
- 03-organisms/by-author/contact-list.twig (this is a working sample.)

> 📋 The template can have a *default value* for the heading level. The *default value* is only used when it’s unable to get the parent’s heading level value. This is the case the second condition might be omitted.

Reviewer: ```suggestion
📋 The template can have a *default value* for the heading level. The *default value* is only used when no heading level value for the parent is present. In this case, the second condition may be omitted.
```

Reviewer: I'm also not sure what the last sentence, about omitting the second condition, means.

[Let's check the heading label next.⬇️](#heading-label)

> **➢ No:** If your heading level number is static or doesn't meet above conditions, you have some work to do💪

Reviewer: ```suggestion
**➢ No:** If your heading level number is static or doesn't meet these conditions, you have some work to do💪
```

> Below info. gives you some idea of what needs to be done.

Reviewer: ```suggestion
Here are some ideas about what you may need to do:
```

> #### Overview
>

> > Only the page title can have `<h1>`.

Reviewer: ```suggestion
> Only the page title should use `<h1>`.
```

>

> > In page templates, the initial heading level is set. Then, before the level changes, the heading level value is adjusted to reflect the content hierarchy in templates.

Reviewer: I'm not quite following this...

Author: This is about passing the heading level value.  Follow up in version 2.

>
> 📋 As of May, 2019, this is managed in Drupal theming, not in Mayflower.  In Mayflower, the heading level value can be passed on.
>
> #### Example
>
> 📄 */patternlab/source/_pattern/01-atoms/04-headings/comp-heading.twig*
>
> The heading level value is passed from its parent template as `compHeading.level`, and assigned to `headingLevel`.
>
> Here the default value is set to `2`.
> 
> ```{% set headingLevel = compHeading.level ? : 2 %}```
>
> The heading level value is ready to render.
>
> ```<h{{ headingLevel }}```
>

<a name="heading-label"></a>
## Heading label

> ### ➤➤ Is the label generic like “Contacts” and “Downloads”?

Reviewer: ```suggestion
### ➤➤ Is the label generic like “Contacts” or “Downloads”?
```

> 🤔 "The label text of/for/about WHAT?" If you don't see the answer for the question, context info is necessary.

Reviewer: ```suggestion
🤔 If the heading label does not answer the question "of/for/about WHAT?" , context information is necessary.
```

> <a name="context"></a>

Reviewer: This isn't rendering visibly -- should it be?

Author: This is not visible.  It's a part of markdown.  It's funny to use HTML like this, but this is how to set up an anchor in a page.  You can do the same trick in markdown as `<a href="#context">` in html with it.

**➣ Yes:** Add context info as visually hidden text for screen reader users.

> 📋 Sighted users can get certain context from the page layout, but screen reader users are usually unable to take advantage of it and don't have the same context. So, we fill the gap by providing the context information as text.

Reviewer: ```suggestion
📋 Sighted users can get certain context from the page layout, but screen reader users don't have the same context. Context info fills the gap for those users.
```

> Some pages could have multiple **Contacts** and **Dowloads** headings in a page. Make sure each one of them is *uniquely* identifiable.
> 
> ##### Examples
> Contacts`<span class=”ma--visually-hidden”>` for the Department of Health office`</span>`
>
> Downloads`<span class=”ma--visually-hidden”>`	 for the marine fishing license`</span>`

**➢ No:** The label is unique. Then, move to the following question.⬇️

### ➤➤ Does the label have enough context to describe the section content?

> 🤔 Check if the label itself would tell what the following content is about. If it doesn't, context info is necessary.

Reviewer: ```suggestion
🤔 Check if the label itself describes the content. If it doesn't, context info is necessary.
```

**➣ Yes:** You are all set with the heading. [Move onto next element.➡️](a11y-checklist.md)

> **➢ No:** Use [above technique](#context) to provide context. ⬆️

Reviewer: ```suggestion
**➢ No:** Use [the techniques above](#context) to provide context. ⬆️
```
